hq_superset/views.py

```diff
@@ -0,0 +1,122 @@
+import pandas as pd
+import superset
+from flask import url_for, render_template, redirect, request, session, g
+from flask_appbuilder import expose, BaseView
+from flask_appbuilder.security.decorators import has_access, permission_name
+from flask_login import current_user
+from io import BytesIO
+from superset import db
+from superset.connectors.sqla.models import SqlaTable
+from superset.sql_parse import Table
+from superset.models.core import Database
+from zipfile import ZipFile
+from .utils import get_datasource_export_url, get_ucr_database
+from .oauth import get_valid_cchq_oauth_token
+
+
+class HQDatasourceView(BaseView):
+
+    def __init__(self):
+        self.route_base = "/a/<domain>/datasource/"
+        super().__init__()
+
+    @expose("/update/<datasource_id>", methods=["POST"])
+    def create_or_update(self, domain, datasource_id):
+        # Fetches data for a datasource from HQ and creates/updates a superset table
+        from .oauth import get_valid_cchq_oauth_token
+        res = refresh_hq_datasource(domain, datasource_id)
+        return res
+
+
+class CCHQApiException(Exception):
+    pass
+
+
+def refresh_hq_datasource(domain, datasource_id):
```

Comment:
Just a thought: This might be easier to write tests for if it's split up into smaller chunks. Maybe something like ...
```python
def refresh_hq_datasource(domain, datasource_id):
    zipfile = fetch_zipfile(domain, datasource_id)
    data_frame = read_csv(zipfile)
    upload_to_table(data_frame)
    return "success"
```

Reply:
Yeah, definitely, there is going to be good amount of refactor once we start writing tests.